{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.10.10","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\n\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","execution":{"iopub.status.busy":"2023-07-16T08:26:02.971173Z","iopub.execute_input":"2023-07-16T08:26:02.971547Z","iopub.status.idle":"2023-07-16T08:26:03.028582Z","shell.execute_reply.started":"2023-07-16T08:26:02.971506Z","shell.execute_reply":"2023-07-16T08:26:03.027446Z"},"trusted":true},"execution_count":1,"outputs":[{"name":"stdout","text":"/kaggle/input/cafa-5-protein-function-prediction/sample_submission.tsv\n/kaggle/input/cafa-5-protein-function-prediction/IA.txt\n/kaggle/input/cafa-5-protein-function-prediction/Test (Targets)/testsuperset.fasta\n/kaggle/input/cafa-5-protein-function-prediction/Test (Targets)/testsuperset-taxon-list.tsv\n/kaggle/input/cafa-5-protein-function-prediction/Train/train_terms.tsv\n/kaggle/input/cafa-5-protein-function-prediction/Train/train_sequences.fasta\n/kaggle/input/cafa-5-protein-function-prediction/Train/train_taxonomy.tsv\n/kaggle/input/cafa-5-protein-function-prediction/Train/go-basic.obo\n/kaggle/input/protbert-embeddings-for-cafa5/train_ids.npy\n/kaggle/input/protbert-embeddings-for-cafa5/train_embeddings.npy\n/kaggle/input/protbert-embeddings-for-cafa5/test_ids.npy\n/kaggle/input/protbert-embeddings-for-cafa5/test_embeddings.npy\n","output_type":"stream"}]},{"cell_type":"code","source":"#import libraiares:\nimport tensorflow as tf\nimport pandas as pd\nimport numpy as np\n\n# we will use bio python to read the sequences\nfrom Bio import SeqIO\n\nimport seaborn as sns\nimport matplotlib.pyplot as plt\n\n\nfrom sklearn.impute import KNNImputer, SimpleImputer\n#import statsmodels.api as sm\n\nfrom sklearn.model_selection import train_test_split, RandomizedSearchCV, GridSearchCV\n#from sklearn.ensemble import RandomForestRegressor\nfrom scipy.stats import spearmanr\nfrom sklearn.metrics import mean_absolute_error\nfrom xgboost import XGBRegressor\nfrom sklearn.preprocessing import StandardScaler\nimport re\nfrom scipy.stats import randint\n\nimport plotly.express as px\nfrom transformers import BertModel, BertTokenizer\nimport torch\n\n# Required for progressbar widget\nimport progressbar\nimport time\nt0start = time.time() \nfrom tqdm import tqdm","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:03.878264Z","iopub.execute_input":"2023-07-16T08:26:03.878689Z","iopub.status.idle":"2023-07-16T08:26:20.543277Z","shell.execute_reply.started":"2023-07-16T08:26:03.878653Z","shell.execute_reply":"2023-07-16T08:26:20.542047Z"},"trusted":true},"execution_count":2,"outputs":[{"name":"stderr","text":"/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:98: UserWarning: unable to load libtensorflow_io_plugins.so: unable to open file: libtensorflow_io_plugins.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so']\ncaused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io_plugins.so: undefined symbol: _ZN3tsl6StatusC1EN10tensorflow5error4CodeESt17basic_string_viewIcSt11char_traitsIcEENS_14SourceLocationE']\n  warnings.warn(f\"unable to load libtensorflow_io_plugins.so: {e}\")\n/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/__init__.py:104: UserWarning: file system plugins are not loaded: unable to open file: libtensorflow_io.so, from paths: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so']\ncaused by: ['/opt/conda/lib/python3.10/site-packages/tensorflow_io/python/ops/libtensorflow_io.so: undefined symbol: _ZTVN10tensorflow13GcsFileSystemE']\n  warnings.warn(f\"file system plugins are not loaded: {e}\")\n","output_type":"stream"}]},{"cell_type":"markdown","source":"# Problem definition:\nThe goal of this competition is to predict the function of a set of proteins.\nThe accurate assignment of biological function to the protein is key to understanding life at the molecular level. However, assigning function to any specific protein can be made difficult due to the multiple functions many proteins have, along with their ability to interact with multiple partners.\nWe are provided with a set of protein sequences on which the participants are asked to predict Gene Ontology (GO) terms in each of the three subontologies: Molecular Function (MF), Biological Process (BP), and Cellular Component (CC). This set of sequences is referred to as test superset.\n# Data:\n1. train_sequences.fasta - amino acid sequences for proteins in training set\n2. testsuperset.fasta - amino acid sequences for proteins on which the predictions should be made\n3. train_terms.tsv - the training set of proteins and corresponding annotated GO terms\n4. train_taxonomy.tsv - taxon ID for proteins in training set\n5. go-basic.obo - ontology graph structure\n6. testsuperset-taxon-list.tsv - taxon ID for proteins in test superset (Note: you may need to use encoding=\"ISO-8859-1\" to read this file in pandas)\n7. IA.txt - Information Accretion for each term. This is used to weight precision and recall (see Evaluation)\n8. sample_submission.csv - a sample submission file in the correct format","metadata":{"execution":{"iopub.status.busy":"2023-06-29T08:12:34.250016Z","iopub.execute_input":"2023-06-29T08:12:34.250375Z","iopub.status.idle":"2023-06-29T08:12:34.261443Z","shell.execute_reply.started":"2023-06-29T08:12:34.250346Z","shell.execute_reply":"2023-06-29T08:12:34.260309Z"}}},{"cell_type":"code","source":"# path to the train and test fasta files\ntrain_fasta = '/kaggle/input/cafa-5-protein-function-prediction/Train/train_sequences.fasta'\ntest_fasta = '/kaggle/input/cafa-5-protein-function-prediction/Test (Targets)/testsuperset.fasta'","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:20.545350Z","iopub.execute_input":"2023-07-16T08:26:20.545672Z","iopub.status.idle":"2023-07-16T08:26:20.550582Z","shell.execute_reply.started":"2023-07-16T08:26:20.545644Z","shell.execute_reply":"2023-07-16T08:26:20.549286Z"},"trusted":true},"execution_count":3,"outputs":[]},{"cell_type":"code","source":"# read train fasta files\ntrain_sequences = SeqIO.parse(train_fasta, 'fasta')","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:20.552537Z","iopub.execute_input":"2023-07-16T08:26:20.553000Z","iopub.status.idle":"2023-07-16T08:26:20.599963Z","shell.execute_reply.started":"2023-07-16T08:26:20.552968Z","shell.execute_reply":"2023-07-16T08:26:20.598941Z"},"trusted":true},"execution_count":4,"outputs":[]},{"cell_type":"code","source":"sequences_dict = SeqIO.to_dict(train_sequences)","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:20.602507Z","iopub.execute_input":"2023-07-16T08:26:20.602832Z","iopub.status.idle":"2023-07-16T08:26:24.815256Z","shell.execute_reply.started":"2023-07-16T08:26:20.602795Z","shell.execute_reply":"2023-07-16T08:26:24.814201Z"},"trusted":true},"execution_count":5,"outputs":[]},{"cell_type":"code","source":"# take a look at the first three instances:\nfirst_three_keys = list(sequences_dict.keys())[:3]\nfirst_three_instances = {key: sequences_dict[key] for key in first_three_keys}\n\n# Printing the first three instances\nfor seq_id, seq_record in first_three_instances.items():\n    print(\"ID:\", seq_id)\n    print(\"Sequence:\", seq_record.seq)\n    print('Description:', seq_record.description)\n    print()","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:24.816964Z","iopub.execute_input":"2023-07-16T08:26:24.817418Z","iopub.status.idle":"2023-07-16T08:26:24.835026Z","shell.execute_reply.started":"2023-07-16T08:26:24.817375Z","shell.execute_reply":"2023-07-16T08:26:24.834167Z"},"trusted":true},"execution_count":6,"outputs":[{"name":"stdout","text":"ID: P20536\nSequence: MNSVTVSHAPYTITYHDDWEPVMSQLVEFYNEVASWLLRDETSPIPDKFFIQLKQPLRNKRVCVCGIDPYPKDGTGVPFESPNFTKKSIKEIASSISRLTGVIDYKGYNLNIIDGVIPWNYYLSCKLGETKSHAIYWDKISKLLLQHITKHVSVLYCLGKTDFSNIRAKLESPVTTIVGYHPAARDRQFEKDRSFEIINVLLELDNKVPINWAQGFIY\nDescription: P20536 sp|P20536|UNG_VACCC Uracil-DNA glycosylase OS=Vaccinia virus (strain Copenhagen) OX=10249 GN=UNG PE=1 SV=1\n\nID: O73864\nSequence: MTEYRNFLLLFITSLSVIYPCTGISWLGLTINGSSVGWNQTHHCKLLDGLVPDQQQLCKRNLELMHSIVRAARLTKSACTSSFSDMRWNWSSIESAPHFTPDLAKGTREAAFVVSLAAAVVSHAIARACASGDLPSCSCAAMPSEQAAPDFRWGGCGDNLRYYGLQMGSAFSDAPMRNRRSGPQDFRLMQLHNNAVGRQVLMDSLEMKCKCHGVSGSCSVKTCWKGLQDISTISADLKSKYLSATKVIPRQIGTRRQLVPREMEVRPVGENELVYLVSSPDYCTQNAKQGSLGTTDRQCNKTASGSESCGLMCCGRGYNAYTEVLVERCQCKYHWCCYVSCKTCKRTVERYVSK\nDescription: O73864 sp|O73864|WNT11_DANRE Protein Wnt-11 OS=Danio rerio OX=7955 GN=wnt11 PE=2 SV=1\n\nID: O95231\nSequence: MRLSSSPPRGPQQLSSFGSVDWLSQSSCSGPTHTPRPADFSLGSLPGPGQTSGAREPPQAVSIKEAAGSSNLPAPERTMAGLSKEPNTLRAPRVRTAFTMEQVRTLEGVFQHHQYLSPLERKRLAREMQLSEVQIKTWFQNRRMKHKRQMQDPQLHSPFSGSLHAPPAFYSTSSGLANGLQLLCPWAPLSGPQALMLPPGSFWGLCQVAQEALASAGASCCGQPLASHPPTPGRPSLGPALSTGPRGLCAMPQTGDAF\nDescription: O95231 sp|O95231|VENTX_HUMAN Homeobox protein VENTX OS=Homo sapiens OX=9606 GN=VENTX PE=1 SV=1\n\n","output_type":"stream"}]},{"cell_type":"code","source":"#create a dataframe of the above fasta files dictionary:\ndata = []\nfor seq_id, seq_record in sequences_dict.items():\n    sequence = str(seq_record.seq)\n    seq_des = str(seq_record.description)\n    data.append({\"ID\": seq_id, \"Sequence\": sequence, \"description\": seq_des})","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:24.836300Z","iopub.execute_input":"2023-07-16T08:26:24.836935Z","iopub.status.idle":"2023-07-16T08:26:25.242388Z","shell.execute_reply.started":"2023-07-16T08:26:24.836903Z","shell.execute_reply":"2023-07-16T08:26:25.241202Z"},"trusted":true},"execution_count":7,"outputs":[]},{"cell_type":"code","source":"df_sequences = pd.DataFrame(data)","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:25.243640Z","iopub.execute_input":"2023-07-16T08:26:25.243997Z","iopub.status.idle":"2023-07-16T08:26:25.422951Z","shell.execute_reply.started":"2023-07-16T08:26:25.243967Z","shell.execute_reply":"2023-07-16T08:26:25.421821Z"},"trusted":true},"execution_count":8,"outputs":[]},{"cell_type":"code","source":"df_sequences.head()","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:25.424489Z","iopub.execute_input":"2023-07-16T08:26:25.425160Z","iopub.status.idle":"2023-07-16T08:26:25.459282Z","shell.execute_reply.started":"2023-07-16T08:26:25.425117Z","shell.execute_reply":"2023-07-16T08:26:25.457970Z"},"trusted":true},"execution_count":9,"outputs":[{"execution_count":9,"output_type":"execute_result","data":{"text/plain":"           ID                                           Sequence  \\\n0      P20536  MNSVTVSHAPYTITYHDDWEPVMSQLVEFYNEVASWLLRDETSPIP...   \n1      O73864  MTEYRNFLLLFITSLSVIYPCTGISWLGLTINGSSVGWNQTHHCKL...   \n2      O95231  MRLSSSPPRGPQQLSSFGSVDWLSQSSCSGPTHTPRPADFSLGSLP...   \n3  A0A0B4J1F4  MGGEAGADGPRGRVKSLGLVFEDESKGCYSSGETVAGHVLLEAAEP...   \n4      P54366  MVETNSPPAGYTLKRSPSDLGEQQQPPRQISRSPGNTAAYHLTTAM...   \n\n                                         description  \n0  P20536 sp|P20536|UNG_VACCC Uracil-DNA glycosyl...  \n1  O73864 sp|O73864|WNT11_DANRE Protein Wnt-11 OS...  \n2  O95231 sp|O95231|VENTX_HUMAN Homeobox protein ...  \n3  A0A0B4J1F4 sp|A0A0B4J1F4|ARRD4_MOUSE Arrestin ...  \n4  P54366 sp|P54366|GSC_DROME Homeobox protein go...  ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>ID</th>\n      <th>Sequence</th>\n      <th>description</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>P20536</td>\n      <td>MNSVTVSHAPYTITYHDDWEPVMSQLVEFYNEVASWLLRDETSPIP...</td>\n      <td>P20536 sp|P20536|UNG_VACCC Uracil-DNA glycosyl...</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>O73864</td>\n      <td>MTEYRNFLLLFITSLSVIYPCTGISWLGLTINGSSVGWNQTHHCKL...</td>\n      <td>O73864 sp|O73864|WNT11_DANRE Protein Wnt-11 OS...</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>O95231</td>\n      <td>MRLSSSPPRGPQQLSSFGSVDWLSQSSCSGPTHTPRPADFSLGSLP...</td>\n      <td>O95231 sp|O95231|VENTX_HUMAN Homeobox protein ...</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>A0A0B4J1F4</td>\n      <td>MGGEAGADGPRGRVKSLGLVFEDESKGCYSSGETVAGHVLLEAAEP...</td>\n      <td>A0A0B4J1F4 sp|A0A0B4J1F4|ARRD4_MOUSE Arrestin ...</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>P54366</td>\n      <td>MVETNSPPAGYTLKRSPSDLGEQQQPPRQISRSPGNTAAYHLTTAM...</td>\n      <td>P54366 sp|P54366|GSC_DROME Homeobox protein go...</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}]},{"cell_type":"code","source":"#create a column with length of sequences:\ndf_sequences['sequence_len'] = df_sequences['Sequence'].apply(len)","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:25.460792Z","iopub.execute_input":"2023-07-16T08:26:25.461206Z","iopub.status.idle":"2023-07-16T08:26:25.568420Z","shell.execute_reply.started":"2023-07-16T08:26:25.461173Z","shell.execute_reply":"2023-07-16T08:26:25.567458Z"},"trusted":true},"execution_count":10,"outputs":[]},{"cell_type":"code","source":"# read test fasta files\ntest_sequences = SeqIO.parse(test_fasta, 'fasta')","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:25.572447Z","iopub.execute_input":"2023-07-16T08:26:25.573055Z","iopub.status.idle":"2023-07-16T08:26:25.578954Z","shell.execute_reply.started":"2023-07-16T08:26:25.573019Z","shell.execute_reply":"2023-07-16T08:26:25.577456Z"},"trusted":true},"execution_count":11,"outputs":[]},{"cell_type":"code","source":"# Remove duplicates based on ID\ntest_sequences = {rec.id: rec for rec in test_sequences}.values()\n\n# Convert the sequences to a dictionary\nsequences_dict_test = SeqIO.to_dict(test_sequences)\n","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:25.582994Z","iopub.execute_input":"2023-07-16T08:26:25.583535Z","iopub.status.idle":"2023-07-16T08:26:29.513838Z","shell.execute_reply.started":"2023-07-16T08:26:25.583484Z","shell.execute_reply":"2023-07-16T08:26:29.512742Z"},"trusted":true},"execution_count":12,"outputs":[]},{"cell_type":"code","source":"#create a dataframe of the above fasta files dictionary:\ndata_test = []\nfor seq_id, seq_record in sequences_dict_test.items():\n    sequence = str(seq_record.seq)\n    seq_des = str(seq_record.description)\n    data_test.append({\"ID\": seq_id, \"Sequence\": sequence, \"description\": seq_des})\n\ndf_sequences_test = pd.DataFrame(data_test)","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:29.515035Z","iopub.execute_input":"2023-07-16T08:26:29.515334Z","iopub.status.idle":"2023-07-16T08:26:30.039602Z","shell.execute_reply.started":"2023-07-16T08:26:29.515308Z","shell.execute_reply":"2023-07-16T08:26:30.038721Z"},"trusted":true},"execution_count":13,"outputs":[]},{"cell_type":"code","source":"#create a column with length of sequences:\ndf_sequences_test['sequence_len'] = df_sequences_test['Sequence'].apply(len)","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:30.041280Z","iopub.execute_input":"2023-07-16T08:26:30.042001Z","iopub.status.idle":"2023-07-16T08:26:30.148512Z","shell.execute_reply.started":"2023-07-16T08:26:30.041954Z","shell.execute_reply":"2023-07-16T08:26:30.147401Z"},"trusted":true},"execution_count":14,"outputs":[]},{"cell_type":"code","source":"df_sequences_test.head()","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:30.149789Z","iopub.execute_input":"2023-07-16T08:26:30.150151Z","iopub.status.idle":"2023-07-16T08:26:30.162355Z","shell.execute_reply.started":"2023-07-16T08:26:30.150122Z","shell.execute_reply":"2023-07-16T08:26:30.161044Z"},"trusted":true},"execution_count":15,"outputs":[{"execution_count":15,"output_type":"execute_result","data":{"text/plain":"       ID                                           Sequence    description  \\\n0  Q9CQV8  MTMDKSELVQKAKLAEQAERYDDMAAAMKAVTEQGHELSNEERNLL...  Q9CQV8\\t10090   \n1  P62259  MDDREDLVYQAKLAEQAERYDEMVESMKKVAGMDVELTVEERNLLS...  P62259\\t10090   \n2  P68510  MGDREQLLQRARLAEQAERYDDMASAMKAVTELNEPLSNEDRNLLS...  P68510\\t10090   \n3  P61982  MVDREQLVQKARLAEQAERYDDMAAAMKNVTELNEPLSNEERNLLS...  P61982\\t10090   \n4  O70456  MERASLIQKAKLAEQAERYEDMAAFMKSAVEKGEELSCEERNLLSV...  O70456\\t10090   \n\n   sequence_len  \n0           246  \n1           255  \n2           246  \n3           247  \n4           248  ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>ID</th>\n      <th>Sequence</th>\n      <th>description</th>\n      <th>sequence_len</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>Q9CQV8</td>\n      <td>MTMDKSELVQKAKLAEQAERYDDMAAAMKAVTEQGHELSNEERNLL...</td>\n      <td>Q9CQV8\\t10090</td>\n      <td>246</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>P62259</td>\n      <td>MDDREDLVYQAKLAEQAERYDEMVESMKKVAGMDVELTVEERNLLS...</td>\n      <td>P62259\\t10090</td>\n      <td>255</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>P68510</td>\n      <td>MGDREQLLQRARLAEQAERYDDMASAMKAVTELNEPLSNEDRNLLS...</td>\n      <td>P68510\\t10090</td>\n      <td>246</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>P61982</td>\n      <td>MVDREQLVQKARLAEQAERYDDMAAAMKNVTELNEPLSNEERNLLS...</td>\n      <td>P61982\\t10090</td>\n      <td>247</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>O70456</td>\n      <td>MERASLIQKAKLAEQAERYEDMAAFMKSAVEKGEELSCEERNLLSV...</td>\n      <td>O70456\\t10090</td>\n      <td>248</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}]},{"cell_type":"code","source":"#Load the terms /label data:\ntrain_terms = pd.read_csv(\"/kaggle/input/cafa-5-protein-function-prediction/Train/train_terms.tsv\",sep=\"\\t\")\nprint(train_terms.shape)","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:30.164386Z","iopub.execute_input":"2023-07-16T08:26:30.164826Z","iopub.status.idle":"2023-07-16T08:26:34.492152Z","shell.execute_reply.started":"2023-07-16T08:26:30.164784Z","shell.execute_reply":"2023-07-16T08:26:34.491036Z"},"trusted":true},"execution_count":16,"outputs":[{"name":"stdout","text":"(5363863, 3)\n","output_type":"stream"}]},{"cell_type":"code","source":"#Load the taxanomy Dataset\ntrain_tax = pd.read_csv(\"/kaggle/input/cafa-5-protein-function-prediction/Train/train_taxonomy.tsv\",sep=\"\\t\")\nprint(train_tax.shape)","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:34.493648Z","iopub.execute_input":"2023-07-16T08:26:34.494037Z","iopub.status.idle":"2023-07-16T08:26:34.630986Z","shell.execute_reply.started":"2023-07-16T08:26:34.494005Z","shell.execute_reply":"2023-07-16T08:26:34.629758Z"},"trusted":true},"execution_count":17,"outputs":[{"name":"stdout","text":"(142246, 2)\n","output_type":"stream"}]},{"cell_type":"markdown","source":" \n # A. Data preprocessing:\n1. check datasets for null values: datasets are free of null values.\n\n2. Create labels for the model:\n\nFirst we will extract all the needed labels(GO term ID) from train_terms.tsv file. There are more than 40,000 labels. In order to simplify our model, we will choose the most frequent 1500 GO term IDs as labels. \n\nThen we will fetch the train_terms data for the relevant labels only. and then we will create the The final labels dataframe (label_df),  composed of 1500 columns and 142246 entries. \n\n3. Create embeddings for the protein sequences. We will be using pretrained embeddings form protBERT to\n","metadata":{}},{"cell_type":"markdown","source":"## 1. check data set for any missing values","metadata":{}},{"cell_type":"code","source":"#check sequences for null values:\ndf_sequences.isna().sum()","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:34.632467Z","iopub.execute_input":"2023-07-16T08:26:34.632799Z","iopub.status.idle":"2023-07-16T08:26:34.799040Z","shell.execute_reply.started":"2023-07-16T08:26:34.632769Z","shell.execute_reply":"2023-07-16T08:26:34.797768Z"},"trusted":true},"execution_count":18,"outputs":[{"execution_count":18,"output_type":"execute_result","data":{"text/plain":"ID              0\nSequence        0\ndescription     0\nsequence_len    0\ndtype: int64"},"metadata":{}}]},{"cell_type":"code","source":"#check sequences for null values:\ndf_sequences_test.isna().sum()","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:34.800648Z","iopub.execute_input":"2023-07-16T08:26:34.801102Z","iopub.status.idle":"2023-07-16T08:26:34.968493Z","shell.execute_reply.started":"2023-07-16T08:26:34.801060Z","shell.execute_reply":"2023-07-16T08:26:34.967646Z"},"trusted":true},"execution_count":19,"outputs":[{"execution_count":19,"output_type":"execute_result","data":{"text/plain":"ID              0\nSequence        0\ndescription     0\nsequence_len    0\ndtype: int64"},"metadata":{}}]},{"cell_type":"code","source":"#check data for null values:\ntrain_terms.isna().sum()","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:34.970033Z","iopub.execute_input":"2023-07-16T08:26:34.970962Z","iopub.status.idle":"2023-07-16T08:26:40.096302Z","shell.execute_reply.started":"2023-07-16T08:26:34.970919Z","shell.execute_reply":"2023-07-16T08:26:40.095077Z"},"trusted":true},"execution_count":20,"outputs":[{"execution_count":20,"output_type":"execute_result","data":{"text/plain":"EntryID    0\nterm       0\naspect     0\ndtype: int64"},"metadata":{}}]},{"cell_type":"code","source":"df_sequences.describe()","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:40.097585Z","iopub.execute_input":"2023-07-16T08:26:40.097941Z","iopub.status.idle":"2023-07-16T08:26:40.128817Z","shell.execute_reply.started":"2023-07-16T08:26:40.097912Z","shell.execute_reply":"2023-07-16T08:26:40.127638Z"},"trusted":true},"execution_count":21,"outputs":[{"execution_count":21,"output_type":"execute_result","data":{"text/plain":"        sequence_len\ncount  142246.000000\nmean      553.636679\nstd       641.728770\nmin         3.000000\n25%       248.000000\n50%       411.000000\n75%       654.000000\nmax     35375.000000","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>sequence_len</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>count</th>\n      <td>142246.000000</td>\n    </tr>\n    <tr>\n      <th>mean</th>\n      <td>553.636679</td>\n    </tr>\n    <tr>\n      <th>std</th>\n      <td>641.728770</td>\n    </tr>\n    <tr>\n      <th>min</th>\n      <td>3.000000</td>\n    </tr>\n    <tr>\n      <th>25%</th>\n      <td>248.000000</td>\n    </tr>\n    <tr>\n      <th>50%</th>\n      <td>411.000000</td>\n    </tr>\n    <tr>\n      <th>75%</th>\n      <td>654.000000</td>\n    </tr>\n    <tr>\n      <th>max</th>\n      <td>35375.000000</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}]},{"cell_type":"markdown","source":"## 2. get labels as label_df\nFirst we will extract all the needed labels(GO term ID) from train_terms.tsv file.\n\nThere are more than 40,000 labels. In order to simplify our model, we will choose the most frequent 1500 GO term IDs as labels.|","metadata":{}},{"cell_type":"code","source":"# We will now save the first 1500 most frequent GO term Ids into a list.\n# Set the limit for label\nnum_of_labels = 1500\n\n# Take value counts in descending order and fetch first 1500 `GO term ID` as labels\nlabels = train_terms['term'].value_counts().index[:num_of_labels].tolist()","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:40:02.366035Z","iopub.execute_input":"2023-07-16T08:40:02.367020Z","iopub.status.idle":"2023-07-16T08:40:03.375346Z","shell.execute_reply.started":"2023-07-16T08:40:02.366921Z","shell.execute_reply":"2023-07-16T08:40:03.374035Z"},"trusted":true},"execution_count":41,"outputs":[]},{"cell_type":"code","source":"# Fetch the train_terms data for the relevant labels only\ntrain_terms_updated = train_terms.loc[train_terms['term'].isin(labels)]\ntrain_terms_updated.shape","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:40:04.164422Z","iopub.execute_input":"2023-07-16T08:40:04.164824Z","iopub.status.idle":"2023-07-16T08:40:04.993611Z","shell.execute_reply.started":"2023-07-16T08:40:04.164793Z","shell.execute_reply":"2023-07-16T08:40:04.992498Z"},"trusted":true},"execution_count":42,"outputs":[{"execution_count":42,"output_type":"execute_result","data":{"text/plain":"(4420737, 3)"},"metadata":{}}]},{"cell_type":"code","source":"train_terms_updated","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:40:09.798382Z","iopub.execute_input":"2023-07-16T08:40:09.798792Z","iopub.status.idle":"2023-07-16T08:40:09.813741Z","shell.execute_reply.started":"2023-07-16T08:40:09.798762Z","shell.execute_reply":"2023-07-16T08:40:09.812421Z"},"trusted":true},"execution_count":43,"outputs":[{"execution_count":43,"output_type":"execute_result","data":{"text/plain":"            EntryID        term aspect\n0        A0A009IHW8  GO:0008152    BPO\n1        A0A009IHW8  GO:0034655    BPO\n3        A0A009IHW8  GO:0044270    BPO\n4        A0A009IHW8  GO:0006753    BPO\n6        A0A009IHW8  GO:0044237    BPO\n...             ...         ...    ...\n5363857      X5L565  GO:0003674    MFO\n5363859      X5L565  GO:0016491    MFO\n5363860      X5M5N0  GO:0005515    MFO\n5363861      X5M5N0  GO:0005488    MFO\n5363862      X5M5N0  GO:0003674    MFO\n\n[4420737 rows x 3 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>EntryID</th>\n      <th>term</th>\n      <th>aspect</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>A0A009IHW8</td>\n      <td>GO:0008152</td>\n      <td>BPO</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>A0A009IHW8</td>\n      <td>GO:0034655</td>\n      <td>BPO</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>A0A009IHW8</td>\n      <td>GO:0044270</td>\n      <td>BPO</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>A0A009IHW8</td>\n      <td>GO:0006753</td>\n      <td>BPO</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>A0A009IHW8</td>\n      <td>GO:0044237</td>\n      <td>BPO</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>5363857</th>\n      <td>X5L565</td>\n      <td>GO:0003674</td>\n      <td>MFO</td>\n    </tr>\n    <tr>\n      <th>5363859</th>\n      <td>X5L565</td>\n      <td>GO:0016491</td>\n      <td>MFO</td>\n    </tr>\n    <tr>\n      <th>5363860</th>\n      <td>X5M5N0</td>\n      <td>GO:0005515</td>\n      <td>MFO</td>\n    </tr>\n    <tr>\n      <th>5363861</th>\n      <td>X5M5N0</td>\n      <td>GO:0005488</td>\n      <td>MFO</td>\n    </tr>\n    <tr>\n      <th>5363862</th>\n      <td>X5M5N0</td>\n      <td>GO:0003674</td>\n      <td>MFO</td>\n    </tr>\n  </tbody>\n</table>\n<p>4420737 rows × 3 columns</p>\n</div>"},"metadata":{}}]},{"cell_type":"code","source":"train_protein_ids = np.load('/kaggle/input/protbert-embeddings-for-cafa5/train_ids.npy')\nprint(train_protein_ids.shape)","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:26:42.018304Z","iopub.execute_input":"2023-07-16T08:26:42.018914Z","iopub.status.idle":"2023-07-16T08:26:42.084354Z","shell.execute_reply.started":"2023-07-16T08:26:42.018874Z","shell.execute_reply":"2023-07-16T08:26:42.083171Z"},"trusted":true},"execution_count":24,"outputs":[{"name":"stdout","text":"(142246,)\n","output_type":"stream"}]},{"cell_type":"code","source":"# Setup progressbar settings.\n# This is strictly for aesthetic.\nbar = progressbar.ProgressBar(maxval=num_of_labels, \\\n    widgets=[progressbar.Bar('=', '[', ']'), ' ', progressbar.Percentage()])\n\n# Create an empty dataframe of required size for storing the labels,\n# i.e, train_size x num_of_labels (1422a46 x 1500)\ntrain_size = train_protein_ids.shape[0] # len(X)\ntrain_labels = np.zeros((train_size ,num_of_labels))\n\n# Convert from numpy to pandas series for better handling\nseries_train_protein_ids = pd.Series(train_protein_ids)\n\n# Loop through each label\nfor i in range(num_of_labels):\n    # For each label, fetch the corresponding train_terms data\n    n_train_terms = train_terms_updated[train_terms_updated['term'] ==  labels[i]]\n    \n    # Fetch all the unique EntryId aka proteins related to the current label(GO term ID)\n    label_related_proteins = n_train_terms['EntryID'].unique()\n    \n    # In the series_train_protein_ids pandas series, if a protein is related\n    # to the current label, then mark it as 1, else 0.\n    # Replace the ith column of train_Y with with that pandas series.\n    train_labels[:,i] =  series_train_protein_ids.isin(label_related_proteins).astype(float)\n    \n    # Progress bar percentage increase\n    bar.update(i+1)\n\n# Notify the end of progress bar \nbar.finish()\n\n# Convert train_Y numpy into pandas dataframe\nlabels_df = pd.DataFrame(data = train_labels, columns = labels)\nprint(labels_df.shape)","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"The final labels dataframe (label_df) is composed of 1500 columns and 142246 entries. We can see all 1500 dimensions(results will be truncated since the number of columns is big) of our dataset by printing out the first 5 entries using the following code:","metadata":{}},{"cell_type":"code","source":"labels_df.head()","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:35:32.528547Z","iopub.status.idle":"2023-07-16T08:35:32.529815Z","shell.execute_reply.started":"2023-07-16T08:35:32.529221Z","shell.execute_reply":"2023-07-16T08:35:32.529270Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Label을 aspect로 바꾸고 BPO, MFO, CCO 로 onhit라벨링 해야하나?","metadata":{}},{"cell_type":"code","source":"# Take value counts in descending order and fetch first 1500 `GO term ID` as labels\nlabels = train_terms['aspect'].value_counts()\nlabels","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:38:50.202976Z","iopub.execute_input":"2023-07-16T08:38:50.203343Z","iopub.status.idle":"2023-07-16T08:38:51.128731Z","shell.execute_reply.started":"2023-07-16T08:38:50.203315Z","shell.execute_reply":"2023-07-16T08:38:51.127165Z"},"trusted":true},"execution_count":36,"outputs":[{"execution_count":36,"output_type":"execute_result","data":{"text/plain":"BPO    3497732\nCCO    1196017\nMFO     670114\nName: aspect, dtype: int64"},"metadata":{}}]},{"cell_type":"code","source":"labels = ['BPO', 'CCO', 'MFO']\nlabels","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:39:17.024414Z","iopub.execute_input":"2023-07-16T08:39:17.024840Z","iopub.status.idle":"2023-07-16T08:39:17.032873Z","shell.execute_reply.started":"2023-07-16T08:39:17.024807Z","shell.execute_reply":"2023-07-16T08:39:17.031634Z"},"trusted":true},"execution_count":38,"outputs":[{"execution_count":38,"output_type":"execute_result","data":{"text/plain":"['BPO', 'CCO', 'MFO']"},"metadata":{}}]},{"cell_type":"code","source":"# Fetch the train_terms data for the relevant labels only\ntrain_terms_updated = train_terms.loc[train_terms['aspect'].isin(labels)]\ntrain_terms_updated.shape","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:39:23.962576Z","iopub.execute_input":"2023-07-16T08:39:23.963017Z","iopub.status.idle":"2023-07-16T08:39:24.729451Z","shell.execute_reply.started":"2023-07-16T08:39:23.962978Z","shell.execute_reply":"2023-07-16T08:39:24.728236Z"},"trusted":true},"execution_count":39,"outputs":[{"execution_count":39,"output_type":"execute_result","data":{"text/plain":"(5363863, 3)"},"metadata":{}}]},{"cell_type":"code","source":"train_terms_updated","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:39:38.516772Z","iopub.execute_input":"2023-07-16T08:39:38.517218Z","iopub.status.idle":"2023-07-16T08:39:38.533136Z","shell.execute_reply.started":"2023-07-16T08:39:38.517185Z","shell.execute_reply":"2023-07-16T08:39:38.531744Z"},"trusted":true},"execution_count":40,"outputs":[{"execution_count":40,"output_type":"execute_result","data":{"text/plain":"            EntryID        term aspect\n0        A0A009IHW8  GO:0008152    BPO\n1        A0A009IHW8  GO:0034655    BPO\n2        A0A009IHW8  GO:0072523    BPO\n3        A0A009IHW8  GO:0044270    BPO\n4        A0A009IHW8  GO:0006753    BPO\n...             ...         ...    ...\n5363858      X5L565  GO:0050649    MFO\n5363859      X5L565  GO:0016491    MFO\n5363860      X5M5N0  GO:0005515    MFO\n5363861      X5M5N0  GO:0005488    MFO\n5363862      X5M5N0  GO:0003674    MFO\n\n[5363863 rows x 3 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>EntryID</th>\n      <th>term</th>\n      <th>aspect</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>A0A009IHW8</td>\n      <td>GO:0008152</td>\n      <td>BPO</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>A0A009IHW8</td>\n      <td>GO:0034655</td>\n      <td>BPO</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>A0A009IHW8</td>\n      <td>GO:0072523</td>\n      <td>BPO</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>A0A009IHW8</td>\n      <td>GO:0044270</td>\n      <td>BPO</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>A0A009IHW8</td>\n      <td>GO:0006753</td>\n      <td>BPO</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>5363858</th>\n      <td>X5L565</td>\n      <td>GO:0050649</td>\n      <td>MFO</td>\n    </tr>\n    <tr>\n      <th>5363859</th>\n      <td>X5L565</td>\n      <td>GO:0016491</td>\n      <td>MFO</td>\n    </tr>\n    <tr>\n      <th>5363860</th>\n      <td>X5M5N0</td>\n      <td>GO:0005515</td>\n      <td>MFO</td>\n    </tr>\n    <tr>\n      <th>5363861</th>\n      <td>X5M5N0</td>\n      <td>GO:0005488</td>\n      <td>MFO</td>\n    </tr>\n    <tr>\n      <th>5363862</th>\n      <td>X5M5N0</td>\n      <td>GO:0003674</td>\n      <td>MFO</td>\n    </tr>\n  </tbody>\n</table>\n<p>5363863 rows × 3 columns</p>\n</div>"},"metadata":{}}]},{"cell_type":"code","source":"train_protein_ids = np.load('/kaggle/input/protbert-embeddings-for-cafa5/train_ids.npy')\ntrain_protein_ids","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:43:38.230181Z","iopub.execute_input":"2023-07-16T08:43:38.230572Z","iopub.status.idle":"2023-07-16T08:43:38.244378Z","shell.execute_reply.started":"2023-07-16T08:43:38.230543Z","shell.execute_reply":"2023-07-16T08:43:38.242891Z"},"trusted":true},"execution_count":47,"outputs":[{"execution_count":47,"output_type":"execute_result","data":{"text/plain":"array(['P20536', 'O73864', 'O95231', ..., 'Q5RGB0', 'A0A2R8QMZ5',\n       'A0A8I6GHU0'], dtype='<U10')"},"metadata":{}}]},{"cell_type":"code","source":"# Setup progressbar settings.\n# This is strictly for aesthetic.\nbar = progressbar.ProgressBar(maxval=num_of_labels, \\\n    widgets=[progressbar.Bar('=', '[', ']'), ' ', progressbar.Percentage()])\n\n# Create an empty dataframe of required size for storing the labels,\n# i.e, train_size x num_of_labels (1422a46 x 3)\ntrain_size = train_protein_ids.shape[0] # len(X)\ntrain_labels = np.zeros((train_size ,num_of_labels))\n\n# Convert from numpy to pandas series for better handling\nseries_train_protein_ids = pd.Series(train_protein_ids)\n\n# Loop through each label\nfor i in range(num_of_labels):\n    # For each label, fetch the corresponding train_terms data\n    n_train_terms = train_terms_updated[train_terms_updated['aspect'] ==  labels[i]]\n    \n    # Fetch all the unique EntryId aka proteins related to the current label(GO term ID)\n    label_related_proteins = n_train_terms['EntryID'].unique()\n    \n    # In the series_train_protein_ids pandas series, if a protein is related\n    # to the current label, then mark it as 1, else 0.\n    # Replace the ith column of train_Y with with that pandas series.\n    train_labels[:,i] =  series_train_protein_ids.isin(label_related_proteins).astype(float)\n    \n    # Progress bar percentage increase\n    bar.update(i+1)\n\n# Notify the end of progress bar \nbar.finish()\n\n# Convert train_Y numpy into pandas dataframe\nlabels_df = pd.DataFrame(data = train_labels, columns = labels)\nprint(labels_df.shape)","metadata":{"execution":{"iopub.status.busy":"2023-07-16T08:45:15.133533Z","iopub.execute_input":"2023-07-16T08:45:15.133955Z","iopub.status.idle":"2023-07-16T08:50:33.984339Z","shell.execute_reply.started":"2023-07-16T08:45:15.133923Z","shell.execute_reply":"2023-07-16T08:50:33.982624Z"},"trusted":true},"execution_count":49,"outputs":[{"name":"stderr","text":"[====================                                                    ]  29%","output_type":"stream"},{"traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)","Cell \u001b[0;32mIn[49], line 17\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[38;5;66;03m# Loop through each label\u001b[39;00m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(num_of_labels):\n\u001b[1;32m     16\u001b[0m     \u001b[38;5;66;03m# For each label, fetch the corresponding train_terms data\u001b[39;00m\n\u001b[0;32m---> 17\u001b[0m     n_train_terms \u001b[38;5;241m=\u001b[39m train_terms_updated[\u001b[43mtrain_terms_updated\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43maspect\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43m  \u001b[49m\u001b[43mlabels\u001b[49m\u001b[43m[\u001b[49m\u001b[43mi\u001b[49m\u001b[43m]\u001b[49m]\n\u001b[1;32m     19\u001b[0m     \u001b[38;5;66;03m# Fetch all the unique EntryId aka proteins related to the current label(GO term ID)\u001b[39;00m\n\u001b[1;32m     20\u001b[0m     label_related_proteins \u001b[38;5;241m=\u001b[39m n_train_terms[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEntryID\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39munique()\n","File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/pandas/core/ops/common.py:72\u001b[0m, in \u001b[0;36m_unpack_zerodim_and_defer.<locals>.new_method\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     68\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mNotImplemented\u001b[39m\n\u001b[1;32m     70\u001b[0m other \u001b[38;5;241m=\u001b[39m item_from_zerodim(other)\n\u001b[0;32m---> 72\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmethod\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mother\u001b[49m\u001b[43m)\u001b[49m\n","File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/pandas/core/arraylike.py:42\u001b[0m, in \u001b[0;36mOpsMixin.__eq__\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     40\u001b[0m \u001b[38;5;129m@unpack_zerodim_and_defer\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__eq__\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     41\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__eq__\u001b[39m(\u001b[38;5;28mself\u001b[39m, other):\n\u001b[0;32m---> 42\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_cmp_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mother\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moperator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meq\u001b[49m\u001b[43m)\u001b[49m\n","File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/pandas/core/series.py:6243\u001b[0m, in \u001b[0;36mSeries._cmp_method\u001b[0;34m(self, other, op)\u001b[0m\n\u001b[1;32m   6240\u001b[0m rvalues \u001b[38;5;241m=\u001b[39m extract_array(other, extract_numpy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, extract_range\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m   6242\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m np\u001b[38;5;241m.\u001b[39merrstate(\u001b[38;5;28mall\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m-> 6243\u001b[0m     res_values \u001b[38;5;241m=\u001b[39m \u001b[43mops\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcomparison_op\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mop\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6245\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_construct_result(res_values, name\u001b[38;5;241m=\u001b[39mres_name)\n","File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/pandas/core/ops/array_ops.py:287\u001b[0m, in \u001b[0;36mcomparison_op\u001b[0;34m(left, right, op)\u001b[0m\n\u001b[1;32m    284\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m invalid_comparison(lvalues, rvalues, op)\n\u001b[1;32m    286\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m is_object_dtype(lvalues\u001b[38;5;241m.\u001b[39mdtype) \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(rvalues, \u001b[38;5;28mstr\u001b[39m):\n\u001b[0;32m--> 287\u001b[0m     res_values \u001b[38;5;241m=\u001b[39m \u001b[43mcomp_method_OBJECT_ARRAY\u001b[49m\u001b[43m(\u001b[49m\u001b[43mop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrvalues\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    289\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    290\u001b[0m     res_values \u001b[38;5;241m=\u001b[39m _na_arithmetic_op(lvalues, rvalues, op, is_cmp\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n","File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/pandas/core/ops/array_ops.py:75\u001b[0m, in \u001b[0;36mcomp_method_OBJECT_ARRAY\u001b[0;34m(op, x, y)\u001b[0m\n\u001b[1;32m     73\u001b[0m     result \u001b[38;5;241m=\u001b[39m libops\u001b[38;5;241m.\u001b[39mvec_compare(x\u001b[38;5;241m.\u001b[39mravel(), y\u001b[38;5;241m.\u001b[39mravel(), op)\n\u001b[1;32m     74\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m---> 75\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mlibops\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mscalar_compare\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mravel\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mop\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     76\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\u001b[38;5;241m.\u001b[39mreshape(x\u001b[38;5;241m.\u001b[39mshape)\n","\u001b[0;31mKeyboardInterrupt\u001b[0m: "],"ename":"KeyboardInterrupt","evalue":"","output_type":"error"}]},{"cell_type":"markdown","source":"## Get the sequence embeddings using protBERT","metadata":{}},{"cell_type":"markdown","source":"### What is ProtBert model?\nPretrained model on protein sequences using a masked language modeling (MLM) objective. ProtBert is based on Bert model which pretrained on a large corpus of protein sequences in a self-supervised fashion. This means it was pretrained on the raw protein sequences only, with no humans labelling them in any way (which is why it can use lots of publicly available data) with an automatic process to generate inputs and labels from those protein sequences.\n\nOne important difference between our Bert model and the original Bert version is the way of dealing with sequences as separate documents. This means the Next sentence prediction is not used, as each sequence is treated as a complete document. The masking follows the original Bert training with randomly masks 15% of the amino acids in the input.\n\nAt the end, the feature extracted from this model revealed that the LM-embeddings from unlabeled data (only protein sequences) captured important biophysical properties governing protein shape. This implied learning some of the grammar of the language of life realized in protein sequences.\n\nref: https://huggingface.co/Rostlab/prot_bert\n\nThe precalculated protein embeddings can be found here: https://www.kaggle.com/datasets/henriupton/protbert-embeddings-for-cafa5","metadata":{}},{"cell_type":"code","source":"%%time\ndevice = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\ntokenizer = BertTokenizer.from_pretrained(\"Rostlab/prot_bert\", do_lower_case=False )\nmodel = BertModel.from_pretrained(\"Rostlab/prot_bert\").to(device)","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:50:19.791593Z","iopub.execute_input":"2023-07-04T10:50:19.791983Z","iopub.status.idle":"2023-07-04T10:50:35.939457Z","shell.execute_reply.started":"2023-07-04T10:50:19.791953Z","shell.execute_reply":"2023-07-04T10:50:35.938137Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Lets take a moment to see how amino acids are tokenized by prot-BERT model","metadata":{}},{"cell_type":"code","source":"#get the token embedding for amino acids\ntoken_embedding = {token: model.get_input_embeddings()(torch.tensor(idx)) for token, idx in tokenizer.get_vocab().items()}","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:50:35.941784Z","iopub.execute_input":"2023-07-04T10:50:35.942376Z","iopub.status.idle":"2023-07-04T10:50:35.96524Z","shell.execute_reply.started":"2023-07-04T10:50:35.942309Z","shell.execute_reply":"2023-07-04T10:50:35.964162Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print(len(token_embedding))","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:50:35.967355Z","iopub.execute_input":"2023-07-04T10:50:35.967744Z","iopub.status.idle":"2023-07-04T10:50:42.061236Z","shell.execute_reply.started":"2023-07-04T10:50:35.967706Z","shell.execute_reply":"2023-07-04T10:50:42.060143Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"#print embedding of aminoacid == ALANINE(A)\nprint(token_embedding['A'])","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:50:42.064656Z","iopub.execute_input":"2023-07-04T10:50:42.065143Z","iopub.status.idle":"2023-07-04T10:50:42.109005Z","shell.execute_reply.started":"2023-07-04T10:50:42.065099Z","shell.execute_reply":"2023-07-04T10:50:42.107889Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print(token_embedding)","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:50:42.110167Z","iopub.execute_input":"2023-07-04T10:50:42.110536Z","iopub.status.idle":"2023-07-04T10:50:42.129478Z","shell.execute_reply.started":"2023-07-04T10:50:42.110506Z","shell.execute_reply":"2023-07-04T10:50:42.128642Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"We will now load the pre calculated protein embeddings created by HENRI UPTON using the Rost Lab's protBERT protein language model.\n\nThe protein embeddings to be used for training are recorded in train_embeddings.npy and the corresponding protein ids are available in train_ids.npy.\n\nFirst, we will load the protein ids of the protein embeddings in the train dataset contained in train_ids.npy into a numpy array.","metadata":{}},{"cell_type":"markdown","source":"The train_protein_ids array consists of 142246 protein_ids. Let us print out the first 5 entries using the following code:","metadata":{}},{"cell_type":"code","source":"train_protein_ids[:5]","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:50:42.130703Z","iopub.execute_input":"2023-07-04T10:50:42.131215Z","iopub.status.idle":"2023-07-04T10:50:42.137476Z","shell.execute_reply.started":"2023-07-04T10:50:42.131185Z","shell.execute_reply":"2023-07-04T10:50:42.136501Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"After loading the files as numpy arrays, we will convert them into Pandas dataframe.\n\nEach protein embedding is a vector of length 1024. We create the resulting dataframe such that there are 1024 columns to represent the values in each of the 1024 places in the vector.","metadata":{}},{"cell_type":"code","source":"#get pretrained embeddings:\n\nimport numpy as np\n# Load the .npy file\ndata_embed = np.load('/kaggle/input/protbert-embeddings-for-cafa5/train_embeddings.npy')","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:50:58.61448Z","iopub.execute_input":"2023-07-04T10:50:58.615301Z","iopub.status.idle":"2023-07-04T10:51:04.616224Z","shell.execute_reply.started":"2023-07-04T10:50:58.615229Z","shell.execute_reply":"2023-07-04T10:51:04.6153Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"#get temp df of the embeddings:\ntrain_df = pd.DataFrame(data_embed)\n#del embeddings_list\ntrain_df .columns = ['Feature_' + str(x) for x in train_df.columns]\ntrain_df ","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:51:33.793827Z","iopub.execute_input":"2023-07-04T10:51:33.794219Z","iopub.status.idle":"2023-07-04T10:51:33.901365Z","shell.execute_reply.started":"2023-07-04T10:51:33.794189Z","shell.execute_reply":"2023-07-04T10:51:33.900255Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"test_embed = np.load('/kaggle/input/protbert-embeddings-for-cafa5/test_embeddings.npy')\n#get temp df for test data embeddings\ntest_df = pd.DataFrame(test_embed)\n#del test_embeddings_list\ntest_df.columns = ['Feature_' + str(x) for x in test_df.columns]\ntest_df","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:51:38.881882Z","iopub.execute_input":"2023-07-04T10:51:38.882298Z","iopub.status.idle":"2023-07-04T10:51:45.426481Z","shell.execute_reply.started":"2023-07-04T10:51:38.882266Z","shell.execute_reply":"2023-07-04T10:51:45.425118Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Model:\nNext, we will use Tensorflow to train a Deep Neural Network with the protein embeddings.","metadata":{}},{"cell_type":"markdown","source":"## 1. define X and y","metadata":{}},{"cell_type":"markdown","source":"## 2. Prepare train-test split","metadata":{}},{"cell_type":"code","source":"# train and validation set\nX = train_df.to_numpy()\ny = labels_df.to_numpy()","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:58:36.982965Z","iopub.execute_input":"2023-07-04T10:58:36.983587Z","iopub.status.idle":"2023-07-04T10:58:36.990299Z","shell.execute_reply.started":"2023-07-04T10:58:36.983544Z","shell.execute_reply":"2023-07-04T10:58:36.988876Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print(X.shape, y.shape)","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:59:15.708315Z","iopub.execute_input":"2023-07-04T10:59:15.709553Z","iopub.status.idle":"2023-07-04T10:59:15.715302Z","shell.execute_reply.started":"2023-07-04T10:59:15.709509Z","shell.execute_reply":"2023-07-04T10:59:15.713982Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)\n\nprint(X_train.shape, X_val.shape)\nprint(y_train.shape, y_val.shape)","metadata":{"execution":{"iopub.status.busy":"2023-07-04T10:59:45.814684Z","iopub.execute_input":"2023-07-04T10:59:45.815099Z","iopub.status.idle":"2023-07-04T10:59:47.324072Z","shell.execute_reply.started":"2023-07-04T10:59:45.815069Z","shell.execute_reply":"2023-07-04T10:59:47.322538Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## 3. ANN model","metadata":{}},{"cell_type":"code","source":"INPUT_SHAPE = [X_train.shape[1]]\nBATCH_SIZE = 5120\n\nmodel = tf.keras.Sequential([\n    tf.keras.layers.BatchNormalization(input_shape=INPUT_SHAPE),    \n    tf.keras.layers.Dense(units=512, activation='relu'),\n    tf.keras.layers.Dense(units=512, activation='relu'),\n    tf.keras.layers.Dense(units=512, activation='relu'),\n    tf.keras.layers.Dense(units=num_of_labels,activation='sigmoid')\n])\n\n\n# Compile model\nmodel.compile(\n    optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),\n    loss='binary_crossentropy',\n    metrics=['binary_accuracy', tf.keras.metrics.AUC()],\n)","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:04:45.949576Z","iopub.execute_input":"2023-07-04T11:04:45.95001Z","iopub.status.idle":"2023-07-04T11:04:46.066119Z","shell.execute_reply.started":"2023-07-04T11:04:45.949977Z","shell.execute_reply":"2023-07-04T11:04:46.064852Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# callbacks\nearlystopping_cb = tf.keras.callbacks.EarlyStopping(monitor=\"loss\", patience=10)\ncheckpoint_cb = tf.keras.callbacks.ModelCheckpoint(\"/kaggle/working/best_model.pro5\", save_best_only=True)","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:05:55.162202Z","iopub.execute_input":"2023-07-04T11:05:55.163318Z","iopub.status.idle":"2023-07-04T11:05:55.16963Z","shell.execute_reply.started":"2023-07-04T11:05:55.163274Z","shell.execute_reply":"2023-07-04T11:05:55.16842Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# train new model\nhistory = model.fit(X_train, y_train, epochs=30,\n                       validation_data=(X_val, y_val),\n                       callbacks=[earlystopping_cb, checkpoint_cb])","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:05:59.442513Z","iopub.execute_input":"2023-07-04T11:05:59.443134Z","iopub.status.idle":"2023-07-04T11:21:07.173Z","shell.execute_reply.started":"2023-07-04T11:05:59.443076Z","shell.execute_reply":"2023-07-04T11:21:07.171715Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Plot the model's loss and accuracy for each epoch\nA validation accuracy of  0.90\n\n\nThe history.history variable contains the information of each epoch.\n\n","metadata":{}},{"cell_type":"code","source":"history_df = pd.DataFrame(history.history)\nhistory_df.loc[:, ['loss']].plot(title=\"Cross-entropy\")\nhistory_df.loc[:, ['binary_accuracy']].plot(title=\"Accuracy\")","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:22:16.37999Z","iopub.execute_input":"2023-07-04T11:22:16.380464Z","iopub.status.idle":"2023-07-04T11:22:17.136984Z","shell.execute_reply.started":"2023-07-04T11:22:16.38043Z","shell.execute_reply":"2023-07-04T11:22:17.134281Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Submission","metadata":{}},{"cell_type":"markdown","source":"Load features for submission","metadata":{}},{"cell_type":"code","source":"%%time\n# fn = '/kaggle/input/protein-embeddings-1/reduced_embeddings_file.npy'\n# fn = '/kaggle/input/protein-embeddings-1/embed_protbert_train_clip_1200_first_70000_prot.csv'\n# fn = '/kaggle/input/t5embeds/train_embeds.npy'\nfn = '/kaggle/input/protbert-embeddings-for-cafa5/test_embeddings.npy'\nprint(fn)\nX_submit = np.load(fn)\nprint(X_submit.shape)\n# X_submit","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:22:35.405729Z","iopub.execute_input":"2023-07-04T11:22:35.406193Z","iopub.status.idle":"2023-07-04T11:22:35.717671Z","shell.execute_reply.started":"2023-07-04T11:22:35.406162Z","shell.execute_reply":"2023-07-04T11:22:35.716482Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Calculate prediction for submission","metadata":{}},{"cell_type":"code","source":"%%time\nY_submit =  model.predict(X_submit)\nprint(Y_submit.shape)","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:22:39.820671Z","iopub.execute_input":"2023-07-04T11:22:39.821529Z","iopub.status.idle":"2023-07-04T11:23:22.084994Z","shell.execute_reply.started":"2023-07-04T11:22:39.821492Z","shell.execute_reply":"2023-07-04T11:23:22.083737Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"%%time \ndf_finalSubmission = pd.DataFrame(columns = ['Protein Id', 'GO Term Id','Prediction'])","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:23:22.08723Z","iopub.execute_input":"2023-07-04T11:23:22.087758Z","iopub.status.idle":"2023-07-04T11:23:22.097629Z","shell.execute_reply.started":"2023-07-04T11:23:22.08772Z","shell.execute_reply":"2023-07-04T11:23:22.096251Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Load protein ids for the submission","metadata":{}},{"cell_type":"code","source":"%%time\nfn = '/kaggle/input/protbert-embeddings-for-cafa5/test_ids.npy'\nvec_test_protein_ids = np.load(fn)\nprint(vec_test_protein_ids.shape)\nvec_test_protein_ids","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:23:23.799206Z","iopub.execute_input":"2023-07-04T11:23:23.800116Z","iopub.status.idle":"2023-07-04T11:23:23.873909Z","shell.execute_reply.started":"2023-07-04T11:23:23.800078Z","shell.execute_reply":"2023-07-04T11:23:23.872789Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"\"Melt\" protein ids","metadata":{}},{"cell_type":"code","source":"%%time \nl = []\nfor k in list(vec_test_protein_ids):\n    l += [ k] * Y_submit.shape[1]\nprint(len(l), l[:20])    \n\ndf_finalSubmission['Protein Id'] = l","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:23:29.703617Z","iopub.execute_input":"2023-07-04T11:23:29.704455Z","iopub.status.idle":"2023-07-04T11:25:10.46966Z","shell.execute_reply.started":"2023-07-04T11:23:29.704419Z","shell.execute_reply":"2023-07-04T11:25:10.468142Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"\"Melt\" Labels (Gene ontology terms )","metadata":{}},{"cell_type":"code","source":"df_finalSubmission['GO Term Id'] = labels * Y_submit.shape[0]\n# df_finalSubmission.head(3)","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:27:09.773985Z","iopub.execute_input":"2023-07-04T11:27:09.774517Z","iopub.status.idle":"2023-07-04T11:27:27.258715Z","shell.execute_reply.started":"2023-07-04T11:27:09.774481Z","shell.execute_reply":"2023-07-04T11:27:27.257661Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Assign predictions","metadata":{}},{"cell_type":"code","source":"df_finalSubmission['Prediction'] = Y_submit.ravel()","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:27:28.837832Z","iopub.execute_input":"2023-07-04T11:27:28.838263Z","iopub.status.idle":"2023-07-04T11:27:30.134595Z","shell.execute_reply.started":"2023-07-04T11:27:28.838227Z","shell.execute_reply":"2023-07-04T11:27:30.133333Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Save","metadata":{}},{"cell_type":"code","source":"%%time\ndf_finalSubmission.to_csv(\"submission.tsv\",header=False, index=False, sep=\"\\t\")","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:32:23.338693Z","iopub.execute_input":"2023-07-04T11:32:23.339194Z","iopub.status.idle":"2023-07-04T11:48:54.617546Z","shell.execute_reply.started":"2023-07-04T11:32:23.339157Z","shell.execute_reply":"2023-07-04T11:48:54.615965Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"df_finalSubmission.head()","metadata":{"execution":{"iopub.status.busy":"2023-07-04T11:48:54.620582Z","iopub.execute_input":"2023-07-04T11:48:54.621368Z","iopub.status.idle":"2023-07-04T11:48:54.632725Z","shell.execute_reply.started":"2023-07-04T11:48:54.621309Z","shell.execute_reply":"2023-07-04T11:48:54.631682Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}